pkg/querier/engine/streaming/operator/selector.go

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operator
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+type Selector struct {
+	Queryable storage.Queryable
+	Start     int64 // Milliseconds since Unix epoch
+	End       int64 // Milliseconds since Unix epoch
+	Timestamp *int64
+	Interval  int64 // In milliseconds
+	Matchers  []*labels.Matcher
+
+	// Set for instant vector selectors, otherwise 0.
+	LookbackDelta time.Duration
+
+	// Set for range vector selectors, otherwise 0.
+	Range time.Duration
+
+	querier                   storage.Querier
+	currentSeriesBatch        *seriesBatch
+	seriesIndexInCurrentBatch int
+}
+
+// There's not too much science behind this number: this is based on the batch size used for chunks streaming.
+const seriesBatchSize = 256
+
+var seriesBatchPool = sync.Pool{New: func() any {
+	return &seriesBatch{
+		series: make([]storage.Series, 0, seriesBatchSize),
+		next:   nil,
+	}
+}}
+
+func (s *Selector) SeriesMetadata(ctx context.Context) ([]SeriesMetadata, error) {
+	if s.currentSeriesBatch != nil {
+		return nil, errors.New("should not call Selector.SeriesMetadata() multiple times")
+	}
+
+	if s.LookbackDelta != 0 && s.Range != 0 {
+		return nil, errors.New("invalid Selector configuration: both LookbackDelta and Range are non-zero")
+	}
+
+	startTimestamp := s.Start
+	endTimestamp := s.End
+
+	if s.Timestamp != nil {
+		startTimestamp = *s.Timestamp
+		endTimestamp = *s.Timestamp
+	}
+
+	rangeMilliseconds := s.Range.Milliseconds()
+	start := startTimestamp - s.LookbackDelta.Milliseconds() - rangeMilliseconds
+
+	hints := &storage.SelectHints{
+		Start: start,
+		End:   endTimestamp,
+		Step:  s.Interval,
+		Range: rangeMilliseconds,
+		// TODO: do we need to include other hints like Func, By, Grouping?
+	}
+
+	var err error
+	s.querier, err = s.Queryable.Querier(start, endTimestamp)
+	if err != nil {
+		return nil, err
+	}
+
+	ss := s.querier.Select(ctx, true, hints, s.Matchers...)
+	s.currentSeriesBatch = seriesBatchPool.Get().(*seriesBatch)
+	incompleteBatch := s.currentSeriesBatch
+	totalSeries := 0
+
+	for ss.Next() {
+		if len(incompleteBatch.series) == cap(incompleteBatch.series) {
```

Comment:
Idea for the future. We could have a `type seriesBatchList` which implements the linked list of `seriesBatch`. Code here may be slightly cleaner.

Reply:
Yep, this is something I want to do, but wanted to get this PR out. It's already on my to-do list for a follow-up PR.